libbeat/outputs/logstash/logstash.go

```diff
@@ -89,113 +66,27 @@ func (lj *logstash) init(cfg *common.Config) error {
-	config *logstashConfig,
-	transp *transport.Config,
-) (mode.ConnectionMode, error) {
-	sendRetries := config.MaxRetries
-	maxAttempts := sendRetries + 1
-	if sendRetries < 0 {
-		maxAttempts = 0
-	}
-
-	settings := modeutil.Settings{
-		Failover:     !config.LoadBalance,
-		MaxAttempts:  maxAttempts,
-		Timeout:      config.Timeout,
-		WaitRetry:    defaultWaitRetry,
-		MaxWaitRetry: defaultMaxWaitRetry,
-	}
+	clients := make([]outputs.NetworkClient, len(hosts))
+	for i, host := range hosts {
+		var client outputs.NetworkClient
 
-	if config.Pipelining == 0 {
-		clients, err := modeutil.MakeClients(cfg, makeClientFactory(config, transp))
+		conn, err := transport.NewClient(transp, "tcp", host, config.Port)
 		if err != nil {
-			return nil, err
+			return outputs.Fail(err)
 		}
-		return modeutil.NewConnectionMode(clients, settings)
-	}
-
-	clients, err := modeutil.MakeAsyncClients(cfg, makeAsyncClientFactory(config, transp))
-	if err != nil {
-		return nil, err
-	}
-	return modeutil.NewAsyncConnectionMode(clients, settings)
-}
 
-func makeClientFactory(
-	cfg *logstashConfig,
-	tcfg *transport.Config,
-) modeutil.ClientFactory {
-	compressLvl := cfg.CompressionLevel
-	maxBulkSz := cfg.BulkMaxSize
-	to := cfg.Timeout
-	ttl := cfg.TTL
-
-	return func(host string) (mode.ProtocolClient, error) {
-		t, err := transport.NewClient(tcfg, "tcp", host, cfg.Port)
-		if err != nil {
-			return nil, err
+		if config.Pipelining > 0 {
+			client, err = newAsyncClient(conn, config)
+		} else {
+			client, err = newSyncClient(conn, config)
 		}
-		return newLumberjackClient(t, compressLvl, maxBulkSz, to, ttl, cfg.Index)
-	}
-}
-
-func makeAsyncClientFactory(
-	cfg *logstashConfig,
-	tcfg *transport.Config,
-) modeutil.AsyncClientFactory {
-	compressLvl := cfg.CompressionLevel
-	maxBulkSz := cfg.BulkMaxSize
-	queueSize := cfg.Pipelining - 1
-	to := cfg.Timeout
-
-	if cfg.TTL != 0 {
-		logp.Warn(`The async Logstash client does not support the "ttl" option`)
-	}
-
-	return func(host string) (mode.AsyncProtocolClient, error) {
-		t, err := transport.NewClient(tcfg, "tcp", host, cfg.Port)
 		if err != nil {
-			return nil, err
+			return outputs.Fail(err)
 		}
-		return newAsyncLumberjackClient(t, queueSize, compressLvl, maxBulkSz, to, cfg.Index)
-	}
-}
-
-func (lj *logstash) Close() error {
-	return lj.mode.Close()
-}
 
-// TODO: update Outputer interface to support multiple events for batch-like
-//       processing (e.g. for filebeat). Batch like processing might reduce
-//       send/receive overhead per event for other implementors too.
-func (lj *logstash) PublishEvent(
-	signaler op.Signaler,
-	opts outputs.Options,
-	data outputs.Data,
-) error {
-	return lj.mode.PublishEvent(signaler, opts, data)
-}
+		client = outputs.WithBackoff(client, config.Backoff.Init, config.Backoff.Max)
+		clients[i] = client
+	}
 
-// BulkPublish implements the BulkOutputer interface pushing a bulk of events
-// via lumberjack.
-func (lj *logstash) BulkPublish(
```

Comment:
yes. I modified the pipeline and output interfaces to be of one kind only. Outputs always get batches and on input it's of pipeline it's always by one event being pushed. This simplifies/removes some extra logic in outputs and pipeline itself for dealing with bulk and non-bulk event handling.